stepsizers.go

```diff
@@ -27,63 +33,85 @@ func (c ConstantStepSize) StepSize(l Location, dir []float64) float64 {
-	// max(MinStepSize, InitialStepFactor / |g|_∞).
+	// Threshold, the step size is estimated by quadratic interpolation,
+	// otherwise it is set to 2*previous step size.
+	// Default value is 1e-12.
+	Threshold float64
+	// The step size at the first iteration is estimated as InitialStepFactor / |g|_∞.
+	// If InitialStepFactor is zero, it will be set to one.
 	InitialStepFactor float64
-	// Minimum step size that is returned.
-	// Default value: 1e-8
+	// The estimated step size is always greater than or equal to MinStepSize.
+	// MinStepSize times GradientAbsTol should always be greater than machine epsilon.
+	// If MinStepSize is zero, it will be set to 1e-3.
 	MinStepSize float64
+	// The estimated step size is always lower than or equal to MaxStepSize.
+	// If MaxStepSize is zero, it will be set to 1.01.
+	MaxStepSize float64
 
-	fPrev float64
+	fPrev        float64
+	dirPrevNorm  float64
+	projGradPrev float64
+	xPrev        []float64
 }
 
-func (q *QuadraticStepSize) Init(l Location, dir []float64) (stepsize float64) {
-	if q.InterpolationCutOff == 0 {
-		q.InterpolationCutOff = 1e-10
+func (q *QuadraticStepSize) Init(l Location, dir []float64) (stepSize float64) {
+	q.xPrev = resize(q.xPrev, len(l.X))
+	if q.Threshold == 0 {
+		q.Threshold = defaultQuadraticThreshold
 	}
 	if q.InitialStepFactor == 0 {
-		q.InitialStepFactor = 1
+		q.InitialStepFactor = defaultInitialStepFactor
 	}
 	if q.MinStepSize == 0 {
-		q.MinStepSize = minStepSize
+		q.MinStepSize = defaultMinimumStepSize
+	}
+	if q.MaxStepSize == 0 {
+		q.MaxStepSize = defaultMaximumStepSize
 	}
 
 	gNorm := floats.Norm(l.Gradient, math.Inf(1))
-	stepsize = math.Max(q.MinStepSize, q.InitialStepFactor/gNorm)
+	stepSize = math.Max(q.MinStepSize, math.Min(q.InitialStepFactor/gNorm, q.MaxStepSize))
 
 	q.fPrev = l.F
-	return stepsize
+	q.dirPrevNorm = floats.Norm(dir, 2)
+	q.projGradPrev = floats.Dot(l.Gradient, dir)
+	copy(q.xPrev, l.X)
+	return stepSize
 }
 
-func (q *QuadraticStepSize) StepSize(l Location, dir []float64) (stepsize float64) {
-	stepsize = 1
-	t := 1.0
-	if !floats.EqualWithinAbsOrRel(l.F, 0, 1e-8, 1e-6) {
-		t = (q.fPrev - l.F) / math.Abs(l.F)
-	}
-	if t > q.InterpolationCutOff {
-		// The relative change between two consecutive function values compared to
-		// the function value itself is large enough, so compute the minimum of
-		// a quadratic interpolant.
-		// Assuming that the received direction is a descent direction,
-		// stepsize will be positive.
-		stepsize = 2 * (l.F - q.fPrev) / floats.Dot(l.Gradient, dir)
-		// Trim the step size to lie in [MinStepSize, 1]
-		stepsize = math.Max(q.MinStepSize, math.Min(1.01*stepsize, 1))
+func (q *QuadraticStepSize) StepSize(l Location, dir []float64) (stepSize float64) {
+	stepSizePrev := floats.Distance(l.X, q.xPrev, 2) / q.dirPrevNorm
+	projGrad := floats.Dot(l.Gradient, dir)
+
+	stepSize = 2 * stepSizePrev
+	if !floats.EqualWithinRel(q.fPrev, l.F, q.Threshold) {
```

Comment:
I am still not convinced. Testing for "the absolute value of the gradient is small" means introducing a new constant (unless we reuse the one from bisection). Testing for "the gradients are opposite" means deciding upon an angle condition between the previous and current gradient and introducing another constant (projected gradients are always negative). All this for an extremely rare situation which this (or any other) step sizer cannot handle well anyway, because the descent directions in this case seem to be unrelated. Moreover, the optimizer has already advanced to the current location in l.X, so if the minimum (along the previous search direction) lies between the previous and current X, it is unfortunate, but irrelevant for the current search direction.

I like this step sizer less and less. CG seems to work slightly better with the step sizer waiting in the other PR, which leaves the steepest descent as the only possible user of this step sizer. However, in steepest descent the consecutive descent directions are not related at all to each other, so theoretically using this step sizer makes no sense in that case, it is just a heuristic.